src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+        } else {
+            linkBeforeElement(value, getNodeByIndex(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeByIndex = getNodeByIndex(index);
+        final T removedElement = nodeByIndex.element;
+        if (nodeByIndex == head) {
+            head = head.next;
+        } else {
+            nodeByIndex.prev.next = nodeByIndex.next;
+        }
+        if (nodeByIndex == tail) {
+            tail = tail.prev;
+        } else {
+            nodeByIndex.next.prev = nodeByIndex.prev;
+        }
+        listSize--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = getCorrectIndex(object);
+        if (index != -1) {
+            remove(index);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return listSize;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return listSize == 0;
+    }
+
+    private void linkFirstElement(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head = newNode;
+        tail = newNode;
+        listSize++;
+    }
+
+    private void linkLastElement(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        tail.next = newNode;
+        tail = newNode;
+        listSize++;
+    }
+
+    private void linkBeforeElement(T value, Node<T> node) {
+        Node<T> prevNode = node.prev;
+        Node<T> newNode = new Node<>(prevNode, value, node);
+        prevNode.next = newNode;
+        node.prev = newNode;
+        listSize++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= listSize) {
+            throw new IndexOutOfBoundsException("Index out of range: " + index);
+        }
+    }
+
+    private void checkIndexToAdd(int index) {
+        if (index < 0 || index > listSize) {
+            throw new IndexOutOfBoundsException("Index out of range: " + index);
+        }
+    }
+
+    private int getCorrectIndex(T element) {
+        for (int i = 0; i < listSize; i++) {
+            Node<T> node = getNodeByIndex(i);
```

Comment:
Better to move from head node by node and check value (cause you have to iterate every time from the begining when you use getNodeByIndex())

Reply:
not fixed 
you have a lot of additional iterations
let's imagine that node, that you need to find locate on `size / 2 + 1` position, how much additional iteration do you need to do to find it?